lib/src/message.dart

```diff
@@ -149,55 +149,94 @@ abstract class Message {
-      newHeaders['content-type'] =
-          'application/octet-stream; charset=${body.encoding.name}';
-    } else {
-      var contentType = new MediaType.parse(newHeaders['content-type'])
-          .change(parameters: {'charset': body.encoding.name});
-      newHeaders['content-type'] = contentType.toString();
+    if (contentType != null) {
+      newHeaders['content-type'] = contentType;
     }
-  }
 
-  if (body.contentLength != null) {
-    var coding = newHeaders['transfer-encoding'];
-    if (coding == null || equalsIgnoreAsciiCase(coding, 'identity')) {
-      newHeaders['content-length'] = body.contentLength.toString();
+    if (contentLength != null) {
+      newHeaders['content-length'] = contentLength;
     }
+
+    return newHeaders;
   }
 
-  return newHeaders;
-}
+  /// Determines the `content-length` from the given [headers] and [body].
+  ///
+  /// Returns the value for the `content-length` header if it should be
+  /// modified, otherwise it returns `null`.
+  static String _contentLengthHeader(Map<String, String> headers, Body body) {
+    var bodyLength = body.contentLength;
+    if (bodyLength == null) return null;
+
+    var contentLengthHeader = bodyLength.toString();
+    if (contentLengthHeader == getHeader(headers, 'content-length'))
+      return null;
+
+    var coding = getHeader(headers, 'transfer-encoding');
 
-/// Returns whether [headers] declares the same encoding as [body].
-bool _sameEncoding(Map<String, String> headers, Body body) {
-  if (body.encoding == null) return true;
+    return coding == null || equalsIgnoreAsciiCase(coding, 'identity')
+        ? contentLengthHeader
+        : null;
+  }
+
+  /// The default media type `application/octet-stream` as defined by HTTP.
+  static final MediaType _defaultMediaType =
+      new MediaType('application', 'octet-stream');
+
+  /// Determines the `content-type` from the given [headers] and [body].
+  ///
+  /// The function looks at the encoding of the body and encoding specified
+  /// within the `content-type` header. The body's encoding will always
+  /// override the value.
+  ///
+  /// Returns the value for the `content-type` header if it should be
+  /// modified, otherwise it returns `null`.
+  static String _contentTypeHeader(Map<String, String> headers, Body body) {
+    var contentTypeHeader = getHeader(headers, 'content-type');
+    var changed = false;
+    MediaType mediaType;
+    Encoding mediaEncoding;
 
-  var contentType = getHeader(headers, 'content-type');
-  if (contentType == null) return false;
+    if (contentTypeHeader != null) {
+      mediaType = new MediaType.parse(contentTypeHeader);
+      mediaEncoding = Encoding.getByName(mediaType.parameters['charset']);
+    } else {
+      mediaType = _defaultMediaType;
+    }
 
-  var charset = new MediaType.parse(contentType).parameters['charset'];
-  return Encoding.getByName(charset) == body.encoding;
+    if ((body.encoding != null) && (body.encoding != mediaEncoding)) {
```

Comment:
We can't split this into

```
if (body.encoding == null) return null;
if (body.encoding != mediaEncoding) return null;
```

The point of the `changed` value is to allow us to insert a `content-type` for `Map` values. This will be in the next PR.

Since `shelf` has no special handling of `Map` types then there would be no `changed` value and the logic would flow like you suggested.

Reply:
Just one thought here. https://github.com/dart-lang/shelf/issues/58 if that is still something you want to have happen then perhaps we should not make map a special case.